torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -443,20 +436,39 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
                 from a call to :meth:`state_dict`
         """
 
+        # Update the state, trusting the ordering in param_groups
+        # Apart from the removal of states not owned by this rank, the pytorch logic is kept
+        # (See torch.optim.optimizer)
+        id_map = {
+            old_id: p
+            for old_id, p in zip(
+                chain.from_iterable((g["params"] for g in state_dict["param_groups"])),
+                chain.from_iterable((g["params"] for g in self.param_groups)),
+            )
+        }
+
+        _param_list = list(chain.from_iterable((g["params"] for g in self.param_groups)))
+
         for key, value in state_dict["state"].items():
-            param = self._index_to_param[key]
+            if key in id_map:
+                param = id_map[key]
 
-            # Populate the sharded optimizer state on the fly
-            if self._param_to_rank[param] != self.rank:
-                state_dict["state"][key] = None
+                # Populate the sharded optimizer state on the fly,
+                # remove the params that this rank does not own
+                if self.param_to_rank[param] != self.rank:
+                    state_dict["state"][key] = {}
+                else:
+                    self.optim.state[param] = _recursive_copy_to_device(value, non_blocking=True, device=param.device)
             else:
+                # Not a param, copied as-is (backward compatibility or exotic optimizers)
+                param = _param_list[key]
                 self.optim.state[param] = _recursive_copy_to_device(value, non_blocking=True, device=param.device)
 
         super().load_state_dict(state_dict)
 
         # Sync with the optimizer param groups
-        ZeroRedundancyOptimizer._sync_param_groups(state_dict["param_groups"], self.param_groups)
```

Comment:
changing that for consistency, sometimes `self.` sometimes `classname.` (since this is a staticmethod)